### PR TITLE
Use single-quoted string for marshalled JSON

### LINKED ISSUE #12641: json module type inline using single-quotes
## Feature request

I noticed that when the JSON is inlined as a double-quoted string, which then requires all double-quotes inside the string to be backslash-escaped. I think this can be improved by using a single-quote string.

For example, this is what I'm seeing in the output for one of my projects:

```javascript
JSON.parse("{\"の\":{\"seq\":\"no\",\"weight\":1918313},\"に\":{\"seq\":\"ni\",\"weight\":1108840},\"た\": ... ");
```

**What is the expected behavior?**

The current behaviour is fine. It seems not to be ideal.

It could probably be something like:

```javascript
JSON.parse('{"の":{"seq":"no","weight":1918313},"に":{"seq":"ni","weight":1108840},"た": ... ');
```

**What is motivation or use case for adding/changing the behavior?**

This will decrease the size of the output since the doublequotes around the JSON object keys will no longer be backslash-escaped.

**How should this be implemented in your opinion?**

Emit the `JSON.parse` argument as a single-quoted string by changing [line 167 of JsonGenerator.js](https://github.com/webpack/webpack/blob/4ab27ff7234ac37f11d29c5e3790c05fd33bfda6/lib/json/JsonGenerator.js#L167) to be:
```javascript
`JSON.parse('${jsonStr}')`
```

instead of:

```javascript
`JSON.parse(${JSON.stringify(jsonStr)})`
```

I'm fairly certain that this change should be well supported on any JS engine (but I'm not _100%_ sure).

**Are you willing to work on this yourself?**

Yes!

### PR DESCRIPTION
**Pull request by david-fong**
Fixes #12641.

**What kind of change does this PR introduce?**

An improvement: Reduced size of emitted inlined JSON.

**Did you add tests for your changes?**

No. Please let me know if it is worth adding tests for a change like this.

**Does this PR introduce a breaking change?**

No.

**What needs to be documented once your changes are merged?**

Nothing. There is no change to any part of the webpack interface.

### CODE REVIEW

**Comment by david-fong** [comment]
I'm not familiar with Jest. I see that I'm getting a snapshot test failure for [this snapshot](https://github.com/webpack/webpack/blob/master/test/__snapshots__/ConfigTestCases.test.js.snap). Is it that that snapshot needs to be updated? I've tested my changes locally with one of my own projects and it behaved as intended.

File: lib/json/JsonGenerator.js [hunk 1]
```diff
@@ -164,7 +164,7 @@ class JsonGenerator extends Generator {
 		const jsonStr = stringifySafe(finalJson);
 		const jsonExpr =
 			jsonStr.length > 20 && typeof finalJson === "object"
-				? `JSON.parse(${JSON.stringify(jsonStr)})`
+				? `JSON.parse('${jsonStr}')`
```
Comment (sokra) [inline lib/json/JsonGenerator.js:167]: This need to escape single quotes in the json. And also `\`.

**Review by sokra — CHANGES_REQUESTED** [review]

**Comment by sokra** [comment]
Yes update the snapshot. Run yarn jest Stats -u

File: lib/json/JsonGenerator.js (hunk 1 above)
Comment (david-fong) [reply lib/json/JsonGenerator.js]: Thanks for catching this! I've added escaping to handle those cases.

File: lib/json/JsonGenerator.js [hunk 2]
```diff
@@ -164,7 +164,7 @@ class JsonGenerator extends Generator {
 		const jsonStr = stringifySafe(finalJson);
 		const jsonExpr =
 			jsonStr.length > 20 && typeof finalJson === "object"
-				? `JSON.parse(${JSON.stringify(jsonStr)})`
+				? `JSON.parse('${jsonStr.replace("\\", "\\\\").replace("'", "\\'")}')`
```
Comment (sokra) [inline lib/json/JsonGenerator.js:167]:
```suggestion
				? `JSON.parse('${jsonStr.replace(/[\\']/g, "\\$&")}')`
```

Comment (sokra) [reply lib/json/JsonGenerator.js:167]: replace() with a string only replaces the first result. You need to use a regexp with g flag. It's also possible to this in a single operation.

**Review by sokra — CHANGES_REQUESTED** [review]

File: lib/json/JsonGenerator.js [hunk 3]
```diff
@@ -164,7 +164,7 @@ class JsonGenerator extends Generator {
 		const jsonStr = stringifySafe(finalJson);
 		const jsonExpr =
 			jsonStr.length > 20 && typeof finalJson === "object"
-				? `JSON.parse(${JSON.stringify(jsonStr)})`
+				? `JSON.parse('${jsonStr.replace(/[\\']/g, "\\$&")}')`
```
Comment (Knagis) [inline lib/json/JsonGenerator.js:167]: newlines should also be escaped as they would break the single quoted string into multiple lines making it invalid.

Comment (Knagis) [reply lib/json/JsonGenerator.js:167]: though stringifySafe does not permit newlines, so that would not happen - but then test is needed to ensure it is not regressed in the future.

Comment (sokra) [reply lib/json/JsonGenerator.js:167]: since this is the output of `JSON.stringify` it won't contain newlines and most other weird stuff is already escaped.

Comment (sokra) [reply lib/json/JsonGenerator.js:167]: We have tests with new lines in JSON, so I guess this should already be covered

**Comment by sokra** [comment]
Thanks

**Comment by zlk89** [comment]
note that this improvement has broken [webpack-subresource-integrity](https://github.com/waysact/webpack-subresource-integrity), since the plugin thinks the contenthash doesn't change, but the actual content has changed.

**Comment by jscheid** [comment]
If your analysis is correct @zlk89 then it's breaking any app using SRI and long-term caching (regardless of whether they're using our plugin or some other way to add SRI) so there's little we can do about it...

**Comment by zlk89** [comment]
@jscheid you are right, it actually is a bug of filename generation on webpack side.

**Comment by sokra** [comment]
@zlk89 I can't reproduce that. Did you disable `optimization.realContentHash`?

**Comment by sokra** [comment]
Please create a repro repo and file a new issue

**Comment by zlk89** [comment]
Just created a repro repo, looks like it is related with the SRI plugin actually https://github.com/waysact/webpack-subresource-integrity/issues/152 @sokra

**Comment by jscheid** [comment]
@zlk89 if I read the description in https://github.com/waysact/webpack-subresource-integrity/issues/152 correctly, you can get different files with the same name, depending on which webpack version you use.

This means that if you use SRI with long-term caching things will break for users who previously loaded one version of the asset and now are loading the other.

The reason is that the SRI hash will have changed, but the file contents will not have changed from the browser's perspective, because it will reuse the old cached value.

In short, SRI will break if you use long-term (content-independent) caching and your file contents change without their names changing.

This is not a limitation in our plugin -- it's a fundamental problem with SRI and caching.

But perhaps I'm missing something?

**Comment by zlk89** [comment]
@jscheid I just updated the description a little bit. So the problem here is: when SRI plugin is used, webpack cannot calculate contenthash correctly, when content is changed. 
As long as I remove the SRI plugin, webpack DOES generate different contenthash filenames for them.

**Comment by jscheid** [comment]
Ah, ok, that's different then. I'll take a look. @sokra off the top of your head, any idea why this might happen?

**Comment by jscheid** [comment]
@sokra I can reproduce this and have tracked it down to `RealContentHashPlugin.getCompilationHooks(compilation).updateHash.tap`. If I tap into it, like you had suggested at one point, the hash for the file containing the JSON code stays the same between 5.21.2 and 5.24.0, even though the contents changed. If I don't tap into it, the file gets a new hash.

I don't think it's anything in the hook code that would cause this, as that's not modifying any of the data passed in, and it looks like the return value (`oldHash`) is also ignored. What gives?

Ref https://github.com/waysact/webpack-subresource-integrity/blob/cc0bc1a693de206512d06f37319bd7df9a6fdd3d/webpack-subresource-integrity/index.ts#L524-L540

**Comment by jscheid** [comment]
Ah, scratch that, it _is_ return a different value in some cases, that must be what's causing it.

**Comment by jscheid** [comment]
No, wait -- it should just be updating SRI values inside assets, it shouldn't have any effect on the final filename hash except indirectly through the hash updates -- but those updates shouldn't negate the effect of the JSON syntax changes?

**Comment by sokra** [comment]
Ah I see. You must not return the oldHash from updateHash. Return undefined when you don't want to customize the hash update.

**Comment by jscheid** [comment]
Ah! Right. Could you update the bindings to allow undefined as a return value? https://github.com/webpack/webpack/blob/3849f45961c60d4bea91b4ef43bb7da7992a0c3a/types.d.ts#L1676
